### PR TITLE
20211001 approvals dialog

### PR DESCRIPTION
**Pull request by jrobbins**
This implements a dialog box that allows API Owners to see and edit their own approval values for various approval fields on a feature.

In this PR:
* Implement the method to fetch details of the feature's existing approval values open the dialog box.
* Rendering methods to display approval values, approval widgets, existing comments, a box to enter a new comment, and save & cancel buttons.
  * Approval fields are only shown if they have at least one value that is in a pending state
  * Unless there are no such fields, in which case all fields are shown
  * There is also a checkbox to show all approval fields
 * Implement handlers for the approval value widget @change event that accumulates a Map of changed values
 * Implement @click handlers for the Save and Cancel buttons.  Save saves the accumulated changes to the server.
 * Add a global CSS rule to make disabled buttons look disabled.

### CODE REVIEW

**Comment by jpmedley** [comment]
I reviewed the small bits of display text and they look good to me.

**Comment by jrobbins** [comment]
This is how the dialog box looks when it is initially opened on a feature that already has some approval values:
image

**Comment by jrobbins** [comment]
@liyangguang Thanks for your review of my previous PR.  That PR and other prequel PRs that already merged should make reviewing this one easier than if it was all in one huge PR.

File: static/elements/chromedash-approvals-dialog.js [hunk 1]
```diff
@@ -1,42 +1,368 @@
...
+            @change=${this.handleSelectChanged}
+          >
+              ${placeholderOption}
+              ${STATE_NAMES.map((valName) => html`
+                <option value="${valName[0]}"
+                  ?selected=${valName[0] == selectedValue}
+                 >${valName[1]}</option>`
+                )}
+            </select>` : html`
+            ${STATE_NAMES[approvalValue.state][1]}
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:196]: Could `STATE_NAMES[approvalValue.state]` be undefined?

File: static/elements/chromedash-approvals-dialog.js [hunk 2]
```diff
@@ -1,42 +1,368 @@
...
+                              numPending < APPROVAL_DEFS.length);
+        this.showAllIntents = numPending == 0;
+        this.changedApprovalsByField = new Map();
+        this.needsSave = false;
+      });
+    const p3 = window.csClient.getComments(this.featureId).then(
+      (res) => {
+        this.comments = res.comments;
+      });
+    Promise.all([p1, p2, p3]).then(() => {
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:151]: Now that this part is async, do we need to show messages like "Loading...", and catch any failed promises and show some error message?
(if the service is reasonably reliable and fast, then maybe it's fine without thoes)

File: static/elements/chromedash-approvals-dialog.js [hunk 3]
```diff
@@ -1,42 +1,368 @@
...
+
+  formatDate(dateStr) {
+    return dateStr.split('.')[0]; // Ignore microseconds.
+  }
+
+  renderApprovalValue(approvalValue) {
+    const selectedValue = (
+      this.changedApprovalsByField.get(approvalValue.field_id) ||
+          approvalValue.state);
+    let placeholderOption = nothing;
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:171]: Readability nit:
```
const placeholderOption = approvalValue.state == -1 ?
    html`<option value="-1" selected>No value</option>
    : nothing;
```

File: static/elements/chromedash-approvals-dialog.js [hunk 4]
```diff
@@ -1,42 +1,368 @@
...
+    for (let fieldId of this.changedApprovalsByField.keys()) {
+      if (this.changedApprovalsByField.get(fieldId) != -1) {
+        promises.push(
+          window.csClient.setApproval(
+            this.feature.id, fieldId,
+            this.changedApprovalsByField.get(fieldId)));
+      }
+    }
+    const commentArea = this.shadowRoot.querySelector('#comment_area');
+    const commentText = commentArea.value;
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:352]: Maybe add .trim().

File: static/elements/chromedash-approvals-dialog.js [hunk 5]
```diff
@@ -1,42 +1,368 @@
...
+    const commentArea = this.shadowRoot.querySelector('#comment_area');
+    const newVal = commentArea.value;
+    if (newVal != '') newNeedsSave = true;
+    for (let fieldId of this.changedApprovalsByField.keys()) {
+      if (this.changedApprovalsByField.get(fieldId) != -1) {
+        newNeedsSave = true;
+      }
+    }
+    this.needsSave = newNeedsSave;
+    console.log(this.needsSave);
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:331]: Debugging code? remove?

File: static/elements/chromedash-approvals-dialog.js [hunk 6]
```diff
@@ -1,42 +1,368 @@
...
-        TODO: dialog content
+        ${this.renderAllApprovals()}
+        ${this.renderAllComments()}
+        ${this.renderControls()}
       </chromedash-dialog>
     `;
   }
+
+  checkNeedsSave() {
+    console.log('In checkNeedsSave()');
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:320]: Debugging code? remove?

File: static/elements/chromedash-approvals-dialog.js [hunk 7]
```diff
@@ -1,42 +1,368 @@
...
+          padding: var(--content-padding);
+          text-align: right;
+        }
+
+        #show_all_checkbox {
+         float: left;
+        }
+
+        textarea {
+          padding: 4px;
```
Comment (liyangguang) [inline static/elements/chromedash-approvals-dialog.js:125]: Maybe add `resize: vertical or none`

**Review by liyangguang — APPROVED** [review]

File: static/elements/chromedash-approvals-dialog.js (hunk 1 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:196]: It should not be.  I have all the enum values for the state covered.

File: static/elements/chromedash-approvals-dialog.js (hunk 5 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:331]: Gone.

File: static/elements/chromedash-approvals-dialog.js (hunk 6 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:320]: Gone.

File: static/elements/chromedash-approvals-dialog.js (hunk 4 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:352]: Added.

File: static/elements/chromedash-approvals-dialog.js (hunk 7 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:125]: I went with `resize: both`. I figure, let the user have the power to do that if they want to.

File: static/elements/chromedash-approvals-dialog.js (hunk 2 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js:161]: Added.

File: static/elements/chromedash-approvals-dialog.js (hunk 3 above)
Comment (jrobbins) [reply static/elements/chromedash-approvals-dialog.js]: Done

**Pull request merged**